File: src/pass/pass.ts
import type { Block } from "../ast/node/block";
import type { Duplex } from "../ast/node/duplex";
import type { Expression } from "../ast/node/expression";
import type { Lambda } from "../ast/node/lambda";
import type { Node } from "../ast/node/node";
import type { Statement } from "../ast/node/statement";
import type { ComponentString, StringComponent } from "../ast/node/string";
import type { Type } from "../ast/node/type";
import type { Value } from "../ast/node/value";
import type { FunctionCall } from "../ast/tree/duplex/functionCall";
import type { Arithmetic } from "../ast/tree/expression/arithmetic";
import type { Function } from "../ast/tree/function/function";
import type { Parameter } from "../ast/tree/function/parameter";
import type { Assignment } from "../ast/tree/statement/assignment";
import { Break } from "../ast/tree/statement/break";
import type { Condition } from "../ast/tree/statement/condition";
import type { Continue } from "../ast/tree/statement/continue";
import type { FunctionDeclaration } from "../ast/tree/statement/functionDeclaration";
import type { Loop } from "../ast/tree/statement/loop";
import type { Return } from "../ast/tree/statement/return";

export abstract class Pass {
    public visit(node: Node<any>) {
        node.visit(this);
    }
    public visitBlock(block: Block) {}
    public visitStatement(statement: Statement) {}
    public visitExpression(expression: Expression<any>) {}
    public visitDuplex(duplex: Duplex<any>) {}
    public visitValue(value: Value<any>) {}
    public visitFunction(func: Function) {}
    public visitParameter(parameter: Parameter) {}
    public visitFunctionDeclaration(declaration: FunctionDeclaration) {}
    public visitForEachLoop(loop: Loop) {}
    public visitFunctionCall(call: FunctionCall) {}
    public visitWhileLoop(loop: Loop) {}
    public visitReturn(ret: Return) {}
    public visitArithmetic(arithmetic: Arithmetic) {}
    public visitContinue(statement: Continue) {}
    public visitBreak(statement: Break) {}
    public visitType(type: Type<any>) {}
    public visitLambda(lambda: Lambda) {}
    public visitCondition(condition: Condition) {}
    public visitAssignment(assignment: Assignment) {}
    public visitComponentString(string: ComponentString) {}
    public visitStringComponent(component: StringComponent<any, any>) {}
}

File: src/main/kotlin/me/honkling/miniscript/pass/Pass.kt
package me.honkling.miniscript.pass

import me.honkling.miniscript.parser.ast.Block
import me.honkling.miniscript.parser.ast.Node
import me.honkling.miniscript.parser.ast.Type
import me.honkling.miniscript.parser.ast.Value
import me.honkling.miniscript.parser.ast.expression.Arithmetic
import me.honkling.miniscript.parser.ast.expression.FunctionCall
import me.honkling.miniscript.parser.ast.expression.ClassDeclaration
import me.honkling.miniscript.parser.ast.function.Function
import me.honkling.miniscript.parser.ast.function.Parameter
import me.honkling.miniscript.parser.ast.prototype.Class
import me.honkling.miniscript.parser.ast.prototype.Field
import me.honkling.miniscript.parser.ast.stack.Environment
import me.honkling.miniscript.parser.ast.statement.Assignment
import me.honkling.miniscript.parser.ast.statement.Break
import me.honkling.miniscript.parser.ast.statement.Continue
import me.honkling.miniscript.parser.ast.statement.FunctionDeclaration
import me.honkling.miniscript.parser.ast.statement.If
import me.honkling.miniscript.parser.ast.statement.Loop
import me.honkling.miniscript.parser.ast.statement.Native
import me.honkling.miniscript.parser.ast.statement.Return

abstract class Pass {
    open fun visit(node: Node<*>) {
        node.accept(this)
    }

    open fun visitEnvironment(environment: Environment) {}
    open fun visitBlock(node: Block) {}
    open fun visitValue(node: Value<*>) {}
    open fun visitType(node: Type<*>) {}
    open fun visitNative(node: Native) {}
    open fun visitLoop(node: Loop) {}
    open fun visitReturn(node: Return) {}
    open fun visitContinue(node: Continue) {}
    open fun visitBreak(node: Break) {}
    open fun visitIf(node: If) {}
    open fun visitAssignment(node: Assignment) {}
    open fun visitFunctionDeclaration(node: FunctionDeclaration) {}
    open fun visitFunctionCall(node: FunctionCall) {}
    open fun visitFunction(node: Function) {}
    open fun visitParameter(node: Parameter) {}
    open fun visitClassDeclaration(node: ClassDeclaration) {}
    open fun visitClass(node: Class) {}
    open fun visitField(node: Field) {}
    open fun visitArithmetic(node: Arithmetic) {}
}